### PR TITLE
CFE-851: Add Dockerfile for upstream coredns

### PR DESCRIPTION
**Pull request by arkadeepsen**
This PR adds the Dockerfile to build an image of the plugin using upstream coredns/coredns.

### CODE REVIEW

File: hack/build-coredns.sh [hunk 1]
```diff
@@ -4,16 +4,22 @@ set -euo pipefail
...
-
-# Add the "ocp_dnsnameresolver" plugin to the cloned openshift/coredns repo.
+if [ ! -z "${1-}" ] && [ "${1}" = "upstream" ]
+then
+    echo "Cloning coredns/coredns repo"
+    git clone https://github.com/coredns/coredns
+else
+    echo "Cloning openshift/coredns repo"
+    git clone https://github.com/openshift/coredns
+fi
```
Comment (alebedev87) [inline hack/build-coredns.sh]:
```suggestion
COREDNS_URL="https://github.com/openshift/coredns"
if [ "${1}" = "upstream" ]; then
    CLONE_URL="https://github.com/coredns/coredns"
done
echo "Cloning from ${COREDNS_URL}"
git clone "${COREDNS_URL}"
```

File: hack/build-coredns.sh [hunk 2]
```diff
@@ -4,16 +4,22 @@ set -euo pipefail
 
 PLUGIN_PATH=$(readlink -f "$(dirname "$0")/..")
 
-# Create a temporary directory for cloning openshift/coredns repo.
+# Create a temporary directory for cloning coredns repo.
```
Comment (alebedev87) [inline hack/build-coredns.sh:13]: Can you please add a comment at the beginning of the script about the arg you can pass and its expected values?

File: Dockerfile.upstream [hunk 3]
```diff
@@ -0,0 +1,14 @@
+FROM golang:1.21 as builder
+WORKDIR /go/src/github.com/openshift/coredns-ocp-dnsnameresolver
+COPY . .
+
+RUN make build-coredns ARGS="upstream"
+
+FROM gcr.io/distroless/base-debian12:latest
```
Comment (alebedev87) [inline Dockerfile.upstream:7]: Any particular reason it's debian based?

Comment (arkadeepsen) [reply Dockerfile.upstream:7]: As upstream coredns uses debian based images, I used the same here. However, I faced some issues with glibc libraries required for go 1.21 when testing the image built using `static-debian`, that's why I used `base-debian`.

File: hack/build-coredns.sh (hunk 2 above)
Comment (arkadeepsen) [reply hack/build-coredns.sh:13]: Added the comment explaining the argument being passed.

File: hack/build-coredns.sh (hunk 1 above)
Comment (arkadeepsen) [reply hack/build-coredns.sh]: Made the required changes to the script.

File: Dockerfile.upstream (hunk 3 above)
Comment (alebedev87) [reply Dockerfile.upstream:7]: Got it, we can have a closer look at the problem with the `static-debian` image later.

**Pull request merged**